test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+    assert.strictEqual(e2, reason);
+    assert.strictEqual(promise2, promise);
+  });
+  var promise2 = new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+    }, 1);
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+    throw e2;
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var e2 = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e2, reason);
+    assert.strictEqual(promise2, promise);
+  });
+  var promise = new Promise(function(_, reject) {
+    setTimeout(function() {
+      reject(e);
+      process.nextTick(function() {
+        promise2 = promise.then(assert.fail, function(reason) {
+          assert.strictEqual(e, reason);
+          throw e2;
+        });
+      });
+    }, 1);
+  });
+  var promise2;
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.reject(e).then(assert.fail, function(){});
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  new Promise(function(_, reject) {
+    reject(e);
+  }).then(assert.fail, function(){});
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = Promise.reject(e);
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = new Promise(function(_, reject) {
+    reject(e);
+  });
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var promise = new Promise(function(_, reject) {
+    reject(e);
+  });
+  process.nextTick(function() {
+    promise.then(assert.fail, function(){});
+  });
+});
+
+// State adapation tests
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
+    return new Promise(function(_, reject) {
+      setTimeout(function() {
+        reject(e);
+      }, 1);
+    });
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
+    throw e;
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
+    return Promise.reject(e);
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var _promise;
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(_promise, promise);
+  });
+  _promise = Promise.resolve().then(function() {
```

Comment:
A rejected promise derived from returning an asynchronously-rejected promise in a fulfillment handler does trigger unhandledRejection